rtt/rtt.go

```diff
@@ -0,0 +1,103 @@
+package rtt
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	errors "github.com/pkg/errors"
+	tcp "github.com/status-im/tcp-shaker"
+)
+
+type Result struct {
+	Address string
+	Latency int
+	PingErr error
+}
+
+func (cr *Result) String() string {
+	return fmt.Sprintf(
+		"Result{address: %s, latency: %d, error: %s}\n",
+		cr.Address, cr.Latency, cr.PingErr)
+}
+
+func runCheck(c *tcp.Checker, address string, ch chan<- Result, timeout time.Duration) {
+	// mesaure RTT
+	start := time.Now()
+	// TCP Ping
+	err := c.CheckAddr(address, timeout)
+	// measure RTT
+	elapsed := time.Since(start)
+	latency := int(elapsed.Milliseconds())
+
+	if err != nil { // don't confuse users with valid latency values on error
+		latency = -1
+		if err == tcp.ErrTimeout {
+			err = errors.Wrap(err, "tcp check timeout")
+		} else {
+			switch err.(type) {
+			case tcp.ErrConnect:
+				err = errors.Wrap(err, "unable to connect")
+			default:
+				err = errors.Wrap(err, "unknown error")
+			}
+		}
+	}
+
+	// return results
+	ch <- Result{
+		Address: address,
+		Latency: latency,
+		PingErr: err,
+	}
+}
+
+func waitForResults(errCh <-chan error, resCh <-chan Result) (results []Result, err error) {
+	for {
+		select {
+		case err = <-errCh:
+			return nil, err
+		case res, ok := <-resCh:
+			if !ok {
+				return
+			}
+			results = append(results, res)
+		}
+	}
+}
+
+func CheckHosts(addresses []string, timeout time.Duration) ([]Result, error) {
+	c := tcp.NewChecker()
+
+	// channel for receiving possible checking loop failure
+	errCh := make(chan error, 1)
+	// channel for returning results from concurrent checks
+	resCh := make(chan Result, len(addresses))
+
+	// stop the checking loop when function exists
+	ctx, stopChecker := context.WithCancel(context.Background())
+	defer stopChecker()
+
+	// loop that queries Epoll and pipes events
+	go func() {
+		// TODO handle error from this?
+		errCh <- c.CheckingLoop(ctx)
+	}()
+
+	var wg sync.WaitGroup
+	for i := 0; i < len(addresses); i++ {
+		wg.Add(1)
+		go func(address string) {
+			defer wg.Done()
+			runCheck(c, address, resCh, timeout)
+		}(addresses[i])
+	}
+	// wait for all the routines to finish
+	wg.Wait()
+	// close the pipe
+	close(resCh)
+
+	// wait for the results for all addresses
+	return waitForResults(errCh, resCh)
```

Comment:
```suggestion
	if err := <-errCh; err != nil {
	    return nil, err
	}
	return <-collectResults(resCh)
```

Or even do:

```go
errCh := make(chan error, 1)
doneChecks := make(chan struct{})
resCh := make(chan Result, len(addresses))

go func() {
		select {
		case err := <-c.CheckingLoop(ctx):
				if err != nil {
					errCh <- err
				}
		case <-doneChecks:
		}
}()

// ...

var result []Result
go func() {
    for item := range resCh {
        result = append(result, item)
    }
}()

// ...

wg.Wait()
close(doneChecks)
close(resCh)

if err := <-errCh; err != nil {
    return nil, err
}
return result, nil
```

It for 99% can be simplified.

Reply:
Adjusted.